opm/autodiff/StandardWell_impl.hpp

```diff
@@ -122,176 +119,59 @@ namespace Opm
-                    // for GRAT controlled wells solvent is included in the target
-                    wellVolumeFractionScaledPhaseUnderControl += wellVolumeFractionScaled(contiSolventEqIdx);
-                }
 
-                if (comp_idx == compIdx_under_control) {
-                    if (has_solvent && compIdx_under_control == FluidSystem::gasCompIdx) {
-                        qs.setValue(target_rate * wellVolumeFractionScaled(compIdx_under_control).value() / wellVolumeFractionScaledPhaseUnderControl.value() );
-                        return qs;
-                    }
-                    qs.setValue(target_rate);
-                    return qs;
-                }
 
-                // TODO: not sure why the single phase under control will have near zero fraction
-                const double eps = 1e-6;
-                if (wellVolumeFractionScaledPhaseUnderControl < eps) {
-                    return qs;
-                }
-                return (target_rate * wellVolumeFractionScaled(comp_idx) / wellVolumeFractionScaledPhaseUnderControl);
-            }
+    template<typename TypeTag>
+    typename StandardWell<TypeTag>::EvalWell
+    StandardWell<TypeTag>::
+    getQs(const int comp_idx) const
+    {
+        // Note: currently, the GTotal definition is still depends on Injector/Producer.
+        assert(comp_idx < num_components_);
 
-            // when it is a combined two phase rate limit, such like LRAT
-            // we neec to calculate the rate for the certain phase
-            if (num_phases_under_rate_control == 2) {
-                EvalWell combined_volume_fraction = 0.;
-                for (int p = 0; p < np; ++p) {
-                    const unsigned compIdxTmp = flowPhaseToEbosCompIdx(p);
-                    if (distr[p] == 1.0) {
-                        combined_volume_fraction += wellVolumeFractionScaled(compIdxTmp);
-                    }
+        if (well_type_ == INJECTOR) { // only single phase injection
+            // TODO: using comp_frac here is dangerous, it should be changed later
+            // Most likely, it should be changed to use distr, or at least, we need to update comp_frac_ based on distr
+            // while solvent might complicate the situation
+            const auto pu = phaseUsage();
+            const int legacyCompIdx = ebosCompIdxToFlowCompIdx(comp_idx);
+            double comp_frac = 0.0;
+            if (has_solvent && comp_idx == contiSolventEqIdx) { // solvent
+                comp_frac = comp_frac_[pu.phase_pos[ Gas ]] * wsolvent();
+            } else if (legacyCompIdx == pu.phase_pos[ Gas ]) {
+                comp_frac = comp_frac_[legacyCompIdx];
+                if (has_solvent) {
+                    comp_frac *= (1.0 - wsolvent());
                 }
-                return (target_rate * wellVolumeFractionScaled(comp_idx) / combined_volume_fraction);
+            } else {
+                comp_frac = comp_frac_[legacyCompIdx];
             }
```

Comment:
Why not just use WellInjector::TypeEnum injectorType from the parser. 
```
If(type == injectorType::Water && comp_idx != Water
 || type == injectorType::Oil && comp_idx != Oil)
||   type == injectorType::Gas && comp_idx != Gas )
   return 0
else 
  return  primary_variables_evaluation_[GTotal]
```

Reply:
I think this is a good opportunity to start getting rid of comp_frac and distr !

Reply:
Removing `comp_frac` and `distr` is better to be put in a separate PR, since it has been used many places and we better limit the scope of the current PR. Especially, we had something unsolved as reported in https://github.com/OPM/opm-simulators/pull/1441 , we need to do that in a separate PR for better testing.

Reply:
```
If(type == injectorType::Water && comp_idx != Water
 || type == injectorType::Oil && comp_idx != Oil)
||   type == injectorType::Gas && comp_idx != Gas )
   return 0
else 
  return  primary_variables_evaluation_[GTotal]
```

solvent part is part I am always a little nervous about. I am not sure how to incorporate solvent related here.